examples/serverless-with-firebase/src/functions/index.js

```diff
@@ -0,0 +1,6 @@
+const functions = require('firebase-functions')
```

Comment:
Do these become completely separate functions? Like separately bundled? Otherwise this wouldn't scale to many pages

Reply:
Yes, this is mainly for demo the idea of using next.js 8 serverless feature. The routing and mapping is decision for the developers to make. For example

/login -> LoginPage function -> login.tsx (1:1)
/UserProfile/** -> UserProfilePage function -> userProfile.tsx (1:n)
**/** -> IndexPage function -> index.tsx (1:m)

Reply:
Also, compare to the following architecture, which create only one cloud function for everything and handle routing internally:

``` javascript
export const nextApp = functions.https.onRequest((req, res) => {
    return app.prepare().then(() => handle(req, res))
})
```

I believe this example would be better since each cloud function only do what is requested by the client.

Reply:
If I understand the change properly, before it was using Express `app.prepare()` and now it's one `page.render()` per function. If that is so, then it respects the "serverless way" much better by avoiding a general handler.

There are pros and cons about both approach, I wonder how middlewares can be applied automatically with this serverless approach, for instance.

Reply:
I actually think the build output should be wrapped and turned into one firebase function per route instead of importing them.

Reply:
I think you might misunderstand the structure here. The `src/functions/index.js` is used for firebase cloud functions only, it is copied to `dist/functions` folder during build and does nothing more. It's isolated from the main next.js app and only import serverless function from next.js for page rendering.
Firebase Cloud Functions use this structure to define multiple functions: https://firebase.google.com/docs/functions/get-started
I believe the code will be splitted during the firebase deployment as it needs to create multiple cloud functions and put write code into each one.

Reply:
@shadowwalker I'm also working on a very similar setup as this, and @timneutkens is spot on in that this does not scale all that well. I'm currently facing that issue and attempting to figure out a path forward. In essence though, your assumption that firebase deploy would split what's packaged based on the functions and their associated requirements is incorrect. When you use a single root `index.js` to deploy multiple functions (per the suggestion of firebase docs) it will deploy each function _but_ underlying will zip the entire source directory (in this case containing all the next pages) and deploy that along side the function. So, for example, if you split the handers into `page1` and `page2` (two google cloud functions) `page1`'s zip will include `page2` and vice versa. Would be curious if @johnapost has run into this if he's going down the same path?

In all honesty I'd much prefer to steer clear of firebase entirely but as my infrastructure choice if google the only way to achieve root (`/` path) and domain mapping is at current via firebase 😞. Ideally if/when google can start to getting parity with AWS lambda and provide this within their own infrastructure then this won't be an issue and a separate framework can be used such as [Serverless](https://serverless.com/) which could totally alleviate this with their per function package include/exclude.

For what it's worth, I threw the idea of a per function include/exclude out over at `firebase-tools` and it was understandably denied: [#1163](https://github.com/firebase/firebase-tools/issues/1163). If I make a choice on a path forwards on solving this I'll definitely try to drop something in here and hope you all can do the same! Or maybe totally prove me wrong on my above assessment which would also be great 😄

Reply:
The approach here differs from the approach by @johnapost that you reference.

Note: Each below are pages in his example app. Took me a moment to get my head around the naming clash.

```
const onRequest = require('firebase-functions').https.onRequest

// These pages are all available in dist once `npm run build:app` is run
const index = require('./next/serverless/pages/index')
const firebase = require('./next/serverless/pages/firebase')
const nextjs = require('./next/serverless/pages/nextjs')
const serverless = require('./next/serverless/pages/serverless')

// This index file serves as a function manifest for Firebase Functions
// Firebase Functions treats each export as its own serverless function
exports.index = onRequest((req, res) => index.render(req, res))
exports.firebase = onRequest((req, res) => firebase.render(req, res))
exports.nextjs = onRequest((req, res) => nextjs.render(req, res))
exports.serverless = onRequest((req, res) => serverless.render(req, res))
```

More here: https://github.com/johnapost/nextjs-firebase-fns-demo/blob/master/src/functions/index.js

Reply:
@wootencl I think you are right about each firebase functions will load with full pack, even the pages not related to the micro service route. However, per the concept of serverless, each micro service should be super light weight regarding to runtime complexity and runtime resource consumption, not the spacing that package files will occupy (disk spacing is relatively cheap this days, but time is always expensive). The serverless feature in next.js 8 reduce the need to figure out routes on next app handler, which save some time on each request.
Here, each page is not that large if you compare to the dependencies in node_modules. Split pages or not, you will have to upload dependencies anyway, which makes the difference not so significant.
As for firebase, I think it's suitable for small teams (1-3 people) for fast prototyping applications, as it provides easy to use library and ready to use common infrastructures (such as authentication). I use to work at AWS, so I tend to think about scale and best practices when I use it. Such as, how to do IAM management and permission control for the future team growth, how to create a CI/CD pipeline in best way, which services are the best choice, how to create CloudFormation stack to follow the infra as code principle...  Again, it's ok to assign tasks to people if you are in a big company, but for me, I am not really going anywhere while thinking about this.

Reply:
@paintedbicycle This footprint is going to the same, unless there is a way to smartly pick related code to each functions and do code splitting.

Reply:
@shadowwalker the very advantage of Next.js is that we split things by entrypoint. Just like your customers don't need to load the entirety of the universe of your functions when they access one page, neither should your servers.